aea/runtime.py

```diff
@@ -253,165 +239,75 @@ def set_loop(self, loop: AbstractEventLoop) -> None:
+        self.multiplexer.start()
+        await self.multiplexer.wait()
 
     async def _start_agent_loop(self) -> None:
         """Start agent main loop asynchronous way."""
         logger.debug("[{}]: Runtime started".format(self._agent.name))
+
+        await self.multiplexer.connection_status.wait(ConnectionStates.connected)
+
         self.task_manager.start()
         if self._decision_maker is not None:  # pragma: nocover
             self.decision_maker.start()
         logger.debug("[{}]: Calling setup method...".format(self._agent.name))
         self._agent.setup()
-        self._state.set(RuntimeStates.running)
         logger.debug("[{}]: Run main loop...".format(self._agent.name))
-        await self.main_loop.run_loop()
-
-    async def _stop_runtime(self) -> None:
-        """
-        Stop runtime.
-
-        Disconnect multiplexer.
-        Tear down agent.
-        Stop agent main loop.
-        """
+        self.main_loop.start()
+        self._state.set(RuntimeStates.running)
         try:
-            if self._async_stop_lock is None:  # pragma: nocover
-                return  # even not started
-
-            async with self._async_stop_lock:
-
-                if self._state.get() in (
-                    RuntimeStates.stopped,
-                    RuntimeStates.stopping,
-                ):  # pragma: nocover
-                    return
-
-                self._state.set(RuntimeStates.stopping)
-                self.main_loop.stop()
-
-                with suppress(BaseException):
-                    await self.main_loop.wait_run_loop_stopped()
-
-                self._teardown()
-
-                await self._multiplexer_disconnect()
-
-        except BaseException:  # pragma: nocover
-            logger.exception("Runtime exception during stop:")
+            await self.main_loop.wait()
+        except asyncio.CancelledError:
+            self.main_loop.stop()
+            await self.main_loop.wait()
             raise
-        finally:
-            self._state.set(RuntimeStates.stopped)
-            logger.debug("[{}]: Runtime stopped".format(self._agent.name))
-
-    def _stop(self) -> None:
-        """
-        Stop synchronously.
-
-        This one calls async functions and does not guarantee to wait till runtime stopped.
-        """
-        logger.debug("Stop runtime coroutine.")
-        if not self.loop.is_running():  # pragma: nocover
-            logger.debug(
-                "Runtime event loop is not running, start loop with `stop` coroutine"
-            )
-
-            with suppress(BaseException):
-                self.loop.run_until_complete(asyncio.sleep(0.01))
-
-            self.loop.run_until_complete(self._stop_runtime())
-            return
-
-        def set_task():
-            self._stopping_task = self.loop.create_task(self._stop_runtime())
 
-        self.loop.call_soon_threadsafe(set_task)
 
-
-class ThreadedRuntime(BaseRuntime):
+class ThreadedRuntime(AsyncRuntime):
```

Comment:
So ThreadedRuntime adds nothing on top of AsyncRuntime..because threading is implemented in Runnable?

Reply:
it's a stub for the moment, will be fixed with next commits